### PR TITLE
Gm geocoder

### PR DESCRIPTION
**Pull request by rafatower**

### CODE REVIEW

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4157/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4158/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4159/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4160/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4178/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4182/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4188/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4190/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4206/))

**Comment by rafatower** [comment]
retest this, please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4208/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4212/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4213/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4217/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4225/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4238/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4253/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4285/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4288/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4304/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4305/))

**Comment by rafatower** [comment]
retest this, please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4312/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4329/))

**Comment by rafatower** [comment]
Things done in this PR:
- Add the `Gme::TableGeocoder` and related classes
- Refactor some legacy code into an `AbstractTableGeocoder` to define a "common interface" (a bit forced but better than the current ducktyping IMHO)
- Add a `TableGeocoderFactory` that should instantiate the right `TableGeocoder` when needed.

Some remarks:
- The central class is `Gme::TableGeocoder`
- It uses a different approach to avoid temporary tables and/or files, as opposed to other geocodes.
- I renamed the existing `google_maps_key` to the more accurate name: `google_maps_query_string` (backwards compatible).
- Some classes are just ports from python code from google, see https://github.com/googlemaps/google-maps-services-python and the following ones:
  - `UrlSigner`
  - `Gme::Client`
 -  `Gme::GeocoderClient`

It still needs more tests but I did not want to defer the review and fix comments.

@juanignaciosl @javisantana can you please review?

File: services/table-geocoder/lib/gme/table_geocoder.rb [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      def update_table(data_block)
+        # At this point, data_block is an Array that looks like this:
+        #   [{:cartodb_id=>1, :searchtext=>"Some real street name", :lat=>19.29544, :lng=>-99.1472101, :cartodb_georef_status=>true},
+        #    {:cartodb_id=>2, :searchtext=>"foo", :cartodb_georef_status=>false}]
+        geocoded = data_block.select {|row| row[:cartodb_georef_status] == true}
+        if geocoded.count > 0
+          geocoded_to_sql = geocoded.map {|row| "(#{row[:cartodb_id]}, #{row[:lng]}, #{row[:lat]})"}.join(',')
+          query_geocoded = %Q{
+            UPDATE #{@qualified_table_name} as target SET
+              the_geom = ST_SetSRID(ST_MakePoint(geocoded.lng,geocoded.lat),4326),
```
Comment (rafatower) [inline services/table-geocoder/lib/gme/table_geocoder.rb]: replace by `CDB_LatLng`?

File: app/models/carto/user.rb [hunk 2]
```diff
@@ -133,6 +133,27 @@ def google_maps_api_key
...
+      read_attribute(:google_maps_private_key)
+    end
+  end
+
+  def google_maps_geocoder_enabled?
+    google_maps_private_key.present? && google_maps_client_id.present?
+  end
+
+  def google_maps_client_id
+    Rack::Utils.parse_nested_query(google_maps_query_string)['client'] if google_maps_query_string
```
Comment (juanignaciosl) [inline app/models/carto/user.rb]: Oh, I didn't know about `parse_nested_query` method, I use [cgi and/or uri parse methods](https://github.com/CartoDB/cartodb/blob/leapfrog_sync_scraping/services/importer/lib/importer/scraping_importer.rb#L34) for this.

File: services/table-geocoder/lib/abstract_table_geocoder.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
+# encoding: utf-8
+
+module CartoDB
+  class AbstractTableGeocoder
```
Comment (juanignaciosl) [inline services/table-geocoder/lib/abstract_table_geocoder.rb]: <3

File: services/table-geocoder/lib/gme/client.rb [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+      def get(endpoint, params, first_request_time=nil, retry_counter=0)
+        first_request_time ||= Time.now
+        elapsed = Time.now - first_request_time
+        if elapsed > @retry_timeout
+          raise Timeout.new('retry timeout expired')
+        end
+
+        if retry_counter > 0
+          delay_seconds = 0.5 * 1.5 ** (retry_counter - 1)
```
Comment (juanignaciosl) [inline services/table-geocoder/lib/gme/client.rb]: This looks like magic numbers to me, could you use constants or add a comment explaining why?

File: services/table-geocoder/lib/gme/client.rb [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        raise Timeout.new('http request timed out') if resp.timed_out?
+
+        if RETRIABLE_STATUSES.include?(resp.code)
+          return self.get(endpoint, params, first_request_time, retry_counter+1)
+        end
+
+        begin
+          get_body(resp)
+        rescue OverQueryLimit
+          Rollbar.report_message('Carto::Gme::Client rescuing from OverQueryLimit exception', 'debug', params)
```
Comment (juanignaciosl) [inline services/table-geocoder/lib/gme/client.rb]: Use `CartoDB.notify_debug` instead. It's just a wrapper right now, but we could add console output easily.

**Comment by juanignaciosl** [comment]
:+1:

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4349/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4353/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4355/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4381/))

**Comment by rafatower** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4405/))

**Pull request merged**